core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java

```diff
@@ -0,0 +1,277 @@
+
+    @Override
+    public Pattern<CorrelatedJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Context context)
+    {
+        // find global aggregation in subquery
+        Optional<AggregationNode> globalAggregation = PlanNodeSearcher.searchFrom(correlatedJoinNode.getSubquery(), context.getLookup())
+                .where(DecorrelateLeftUnnestWithGlobalAggregation::isGlobalAggregation)
+                .recurseOnlyWhen(node -> node instanceof ProjectNode || isGroupedAggregation(node))
+                .findFirst();
+
+        if (globalAggregation.isEmpty()) {
+            return Result.empty();
+        }
+
+        // find unnest in subquery
+        Optional<UnnestNode> subqueryUnnest = PlanNodeSearcher.searchFrom(correlatedJoinNode.getSubquery(), context.getLookup())
+                .where(node -> isSupportedUnnest(node, correlatedJoinNode.getCorrelation(), context.getLookup()))
+                .recurseOnlyWhen(node -> node instanceof ProjectNode || isGlobalAggregation(node) || isGroupedAggregation(node))
+                .findFirst();
+
+        if (subqueryUnnest.isEmpty()) {
+            return Result.empty();
+        }
+
+        UnnestNode unnestNode = subqueryUnnest.get();
+
+        // assign unique id to input rows to restore semantics of aggregations after rewrite
+        PlanNode input = new AssignUniqueId(
+                context.getIdAllocator().getNextId(),
+                correlatedJoinNode.getInput(),
+                context.getSymbolAllocator().newSymbol("unique", BIGINT));
+
+        // pre-project unnest symbols if they were pre-projected in subquery
+        // The correlated UnnestNode either unnests correlation symbols directly, or unnests symbols produced by a projection that uses only correlation symbols.
+        // Here, any underlying projection that was a source of the correlated UnnestNode, is appended as a source of the rewritten UnnestNode.
+        // If the projection is not necessary for UnnestNode (i.e. it does not produce any unnest symbols), it should be pruned afterwards.
+        PlanNode unnestSource = context.getLookup().resolve(unnestNode.getSource());
+        if (unnestSource instanceof ProjectNode) {
+            ProjectNode sourceProjection = (ProjectNode) unnestSource;
+            input = new ProjectNode(
+                    sourceProjection.getId(),
+                    input,
+                    Assignments.builder()
+                            .putIdentities(input.getOutputSymbols())
+                            .putAll(sourceProjection.getAssignments())
+                            .build());
+        }
+
+        // rewrite correlated join to UnnestNode
+        UnnestNode rewrittenUnnest = new UnnestNode(
+                context.getIdAllocator().getNextId(),
```

Comment:
nit: Maybe you can reuse `unnestNode.getId` here and then there is no need to pass `PlanNodeId correlatedUnnestId` to `rewriteNodeSequence`

Reply:
I can reuse the `PlanNodeId`, but I think that it should be the `CorrelatedJoinNode`'s id rather than `UnnestNode`'s id.
Anyway, I think that it would still be necessary to pass `PlanNodeId correlatedUnnestId`. Could you explain how I could avoid that?

Reply:
If we reused `UnnestNode` id it would not be necessary. As you could do:
```java
        // rewrite correlated join to UnnestNode
        UnnestNode rewrittenUnnest = new UnnestNode(
                unnestNode.getId(),
                input,
                input.getOutputSymbols(),
                unnestNode.getMappings(),
                unnestNode.getOrdinalitySymbol(),
                LEFT,
                Optional.empty());

        // restore all projections, grouped aggregations and global aggregations from the subquery
        PlanNode result = rewriteNodeSequence(
                context.getLookup().resolve(correlatedJoinNode.getSubquery()),
                input.getOutputSymbols(),
                rewrittenUnnest,
                context.getLookup());
```

and then
```java
    private static PlanNode rewriteNodeSequence(PlanNode root, List<Symbol> leftOutputs, PlanNode sequenceSource, Lookup lookup)
    {
        // bottom of sequence reached -- attach the rewritten source node
        if (root.getId().equals(sequenceSource.getId())) {
            return sequenceSource;
        }
```

But it is not anything super important.

Reply:
Right. But I think it might be confusing: having two nodes with the same id at the same time.
And also, `rewrittenUnnest` is the logical successor of `correlatedJoinNode` rather than of `unnestNode`.

Reply:
I'm leaning towards not reusing PlanNodeId here at all.

Reply:
Sure. Keep it.